Restrict function pointers by name

danpoe opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->
Replaces #5174.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue committed:
Add restrict-function-pointer-by-name option

This works similar to restrict-function-pointer, but for names of individual
function pointer variables (globals, locals, parameters) rather than call sites.
This isn't applicable to all situations (for example, calling function pointers
in structs or function pointers returned from functions), but is more readily
applicable to some common use scenarios (e.g. global function pointers loaded at
start time like in OpenGL).

hannes-steffenhagen-diffblue commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -258,21 +271,21 @@ parse_function_pointer_restrictions_from_command_line(
       throw invalid_command_line_argument_exceptiont{
         "couldn't find target name before '/' in `" + restriction_opt + "'",
-        "--" RESTRICT_FUNCTION_POINTER_OPT};
+        option_name};
```
:no_entry_sign: `option`

hannes-steffenhagen-diffblue commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -328,22 +349,83 @@ parse_function_pointer_restrictions_from_file(
   return merged_restrictions;
 }
+
+function_pointer_restrictionst::restrictionst
```
The purpose of this needs to be documented, I am fairly sure I wrote this and it took me like 5 minutes trying to figure out what this was doing and why

danpoe committed:
Pass command line option to parse_function_pointer_restriction()

danpoe committed:
Refactorings and cleanups for the restrict function pointer by name feature

danpoe committed:
Update restrict function pointer by name regression tests

danpoe committed:
Add unit tests for the restrict function pointers by name feature

thk123 commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1037,13 +1037,20 @@ void goto_instrument_parse_optionst::instrument_goto_program()
+
+      const auto function_pointer_restrictions =
+        by_name_restrictions.merge(function_pointer_restrictionst::from_options(
+          options, log.get_message_handler()));
```
```suggestion
      const auto by_callsite_restrictions = function_pointer_restrictionst::from_options(options, log.get_message_handler());
      const auto function_pointer_restrictions =  by_name_restrictions.merge(by_callsite_restrictions);
```

Just to make it clear what the two things that are being merged are.

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -459,3 +466,61 @@ void function_pointer_restrictionst::write_to_file(
 
   function_pointer_restrictions_json.output(outFile);
 }
+function_pointer_restrictionst function_pointer_restrictionst::merge(
```
:bulb: Given that is const, I think a free function that takes two and returns result would be clearer

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -104,12 +104,13 @@ source_locationt make_function_pointer_restriction_assertion_source_location(
-template <typename Handler>
-void for_each_function_call(goto_functiont &goto_function, Handler handler)
+template <typename Handler, typename GotoFunctionT>
+void for_each_function_call(GotoFunctionT &&goto_function, Handler handler)
```
:question: Why does this become generic?

danpoe replied:
This is so that the function can take both const and non-const goto function objects.

thk123 replied:
I don't follow - you can pass a non-const version of something to a function that takes a const object? I'll try undoing this change locally and see if then it clicks!

thk123 replied:
Oh presumably because the handler might require a non-const version.

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -244,9 +248,11 @@ function_pointer_restrictionst::merge_function_pointer_restrictions(
+function_pointer_restrictionst::
+parse_function_pointer_restrictions_from_command_line(
+  const std::list<std::string> &restriction_opts,
+  const std::string &option_name)
```
:no_entry_sign: This variable doesn't appear to be used

thk123 replied:
I guess it should be used on line 274

thk123 replied:
Though I think it would be better to have this throw an exception `invalid_restrictionst` or something, and have the callsite catch it and throw it as a command line exception, as this code doesn't need to know that it is coming from the command line or whatevre

danpoe replied:
Good idea. I've reworked the error handling to now throw an `invalid_restriction_exceptiont`.

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -459,3 +466,61 @@ void function_pointer_restrictionst::write_to_file(
+
+function_pointer_restrictionst get_function_pointer_by_name_restrictions(
+  const goto_modelt &goto_model,
+  const optionst &options)
```
:bulb: I think it would be easier to see what is going on if you have this take the result of ` options.get_list_option(RESTRICT_FUNCTION_POINTER_BY_NAME_OPT)` to decouple the CLI from the parsing of the lsit

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -459,3 +466,61 @@ void function_pointer_restrictionst::write_to_file(
+    for_each_function_call(
+      goto_function.second, [&](goto_programt::const_targett location) {
+        PRECONDITION(location->is_function_call());
+        if(can_cast_expr<dereference_exprt>(
```
:bulb: Use early return to avoid nesting
:bulb: :bulb: Use make_range & filter to express this as a series of transformations which breaks down the loop into understandable chunks

danpoe replied:
Pulled out the loop body into a separate method and using early return.

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -459,3 +466,61 @@ void function_pointer_restrictionst::write_to_file(
+          auto const &function_pointer_call_site = to_symbol_expr(
+            to_dereference_expr(location->get_function_call().function())
+              .pointer());
+          auto const &body = goto_function.second.body;
```
This block of code is hard for me to understand - perhaps some named helper functions would make it clear what is going on here

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -459,3 +466,61 @@ void function_pointer_restrictionst::write_to_file(
+            to_dereference_expr(location->get_function_call().function())
+              .pointer());
+          auto const &body = goto_function.second.body;
+          for(auto it = std::prev(location); it != body.instructions.end();
```
:question: Wait what? We start before the location, and then walk forward looking for an assignment? Shouldn't we be going backwards?

danpoe replied:
Ah yes. The loop is unnecessary as the instruction we look for will always be at `std::prev(location)`. Have removed the loop.

thk123 commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -1038,18 +1038,15 @@ void goto_instrument_parse_optionst::instrument_goto_program()
     parse_function_pointer_restriction_options_from_cmdline(cmdline, options);
 
     if(
-      !options.get_list_option(RESTRICT_FUNCTION_POINTER_OPT).empty() ||
```
:pick: please don't but a bunch of unrelated refactors into the same commit as it makes it harder to review

thk123 replied:
And means you can't explain each one in the commit message

danpoe replied:
Agreed. Will try to be more careful in the future.

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -193,6 +198,51 @@ void handle_call(
   }
 }
+
+void get_by_name_restriction(
```
:no_entry_sign: I think it is suprising that a function called `get_...` doesn't return anything, can this be structured in a way so that it actually does get something? Otherwise, a comment would help

thk123 replied:
Yeah I think it'd be much clearer to return a new restrictions (and append at the callsite if required)

danpoe replied:
The function now returns an `optionalt<restrictiont>`.

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -193,6 +198,51 @@ void handle_call(
+
+  auto const &function_pointer_call_site = to_symbol_expr(
+    to_dereference_expr(function)
+      .pointer());
```
:question: Is it guaranteed that the `pointer()` result will be a symbol expr? I think it either needs a PRECONDITION or an early return as this isn't obvious to me

danpoe replied:
Yes, this is guaranteed due to function pointer labelling having been run before. I think the precondition is redundant here as `to_symbol_expr()` (and all other such conversion functions) already include a precondition as their first statement.

thk123 replied:
Fair - however, if I'm debugging this and I get an error saying isn't a symbol I don't have any clue why you'd expect this, whereas if you provide a specific one, you can add a comment saying must be symbol because pointer_labelling has been run (and therefore, hitting this suggests someone has forgot to run pointer labelling).

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -193,6 +198,51 @@ void handle_call(
+      restrictions.emplace(
+        function_pointer_call_site.get_identifier(),
+        restriction->second);
+    }
```
My understanding is if we have

```
(*fp)()
```

Then we look for lines that look like
```
fp = <something>
```

On the proceeding line, or any line after.

I still think we should be searching backwards, but this still seems off. What if the call is on the actual function we have a named restriction for? I realize from docs that this is a simple case data flow analysis, but is there not already some code that does a data flow analysis that provides this?

danpoe replied:
The loop was actually unnecessary as the assignment will always be on the previous line. I've removed the loop.

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -24,6 +24,11 @@ void typecheck_function_pointer_restrictions(
+  const std::string options =
+    "--" RESTRICT_FUNCTION_POINTER_OPT "/"
+    "--" RESTRICT_FUNCTION_POINTER_FROM_FILE_OPT "/"
+    "--" RESTRICT_FUNCTION_POINTER_BY_NAME_OPT;
```
:question: Is it not possible to resitrct by name from file?

thk123 replied:
Since this is constant - perhaps worth declaring as global constant and then not having to pass into each of the functions

danpoe replied:
Currently that's not supported as the from file option was mostly intended for consuming the output from goto-analyzer which will refer to the calls via indices rather than names. Though it would also make sense to let a user supply manually written json files at some point.

danpoe replied:
The string is gone now since the error handling has changed.

thk123 reviewed (commented):
Will review tests after lunch. 

Note to other reviews: it is easier to review all at once rather than by commit

thk123 commented on regression/goto-instrument/restrict-function-pointer-by-name-global/test.c:
```diff
@@ -0,0 +1,28 @@
+void main()
+{
+  int cond;
+  fp = cond ? f : g;
```
:question: In my experience, cbmc can be smart enough to spot that `h` is unused, I guess the most precise way to test this is fixed, is have a `test-no-restriction.desc` that is the same, but check that the verification fails on the assertion :sheep:

danpoe replied:
cbmc should always be able to figure out which functions are really called, so the verification result without using `--restrict-function-pointer` would be the same. Just the `remove_function_pointers()` pass would probably produce a superset of the functions that a user might specify via `--restrict-function-pointer`. For testing `--restrict-function-pointer`, I think it does not matter what `remove_function_pointers()` might do, as the user could also specify a superset of the functions identified by that pass.

To check that really the correct case distinction is produced, I've now also added regexes that check that precisely the correct functions are called in the transformed goto program.

thk123 commented on regression/goto-instrument/restrict-function-pointer-by-name-local/test.c:
```diff
@@ -0,0 +1,19 @@
+
+void main()
+{
+  fp_t fp = f;
```
:sheep:

thk123 commented on regression/goto-instrument/restrict-function-pointer-to-multiple-functions-via-file-and-command-line-options/test.c:
```diff
@@ -35,38 +19,19 @@ int g(int x)
+  use_f(other);
 }
 
-fptr_t get_f(void)
```
I guess all this deletion would be clearer if put in its own commit explaining why

thk123 commented on regression/goto-instrument/restrict-function-pointer-by-name-local/test.desc:
```diff
@@ -0,0 +1,9 @@
+^EXIT=0$
+^SIGNAL=0$
+--
+--
```
:no_entry_sign: please make sure _all_ tests have an explanation of what they are supposed to be testing. The name is necessarily terse, and in 6 months we definitely will get confused.

danpoe replied:
Added

thk123 commented on unit/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -88,6 +95,67 @@ void merge_restrictions_test()
+
+  label_function_pointer_call_sites(goto_model);
+
+  const auto restrictions =
```
:bulb: Ideally a unit test would test less components at once (e.g. this tests both the parsing of the options, and the restriction stuff) - if you split up the parsing, and have the result of the parsing being passed into the restrictions, the test would be simpler

thk123 replied:
Now I understand, I think the test would have been easier to follow if was a couple of simpler examples:


```
void func1(fp_t fp)
    {
      fp();
    }
void func1(fp_t fp)
    {
      fp();
    }

TEST_CASE("Check function scope is correctly applied to variable names")
fp_restrictionst::get_function_pointer_by_name_restrictions(
      {"func1::fp/fa", "func1::fp/fa"},
      goto_model);
 const auto set1 = restrictions.at("func.function_pointer_call.1");
REQUIRE(restrictions == {{"func2.function_pointer_call.1", "fa"}, {"func2.function_pointer_call.1", "fb"})
...
``

then have a separate test case for checking the scope thing

danpoe replied:
Split it up into several simpler cases.

thk123 replied:
Much clearer thanks!

thk123 commented on unit/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -148,3 +216,8 @@ TEST_CASE("Json conversion", "[core]")
     function_pointer_restrictions2.restrictions);
 }
+
+TEST_CASE("Get function pointer by name restrictions", "[core]")
```
:no_entry_sign: The test should be tagged with `[goto-programs][restrict-function-pointers]`

danpoe replied:
Done

thk123 commented on unit/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -148,3 +216,8 @@ TEST_CASE("Json conversion", "[core]")
+
+TEST_CASE("Get function pointer by name restrictions", "[core]")
+{
+  get_function_pointer_by_name_restrictions_test();
```
:grey_question: Why not just put the function in the test? Doing so makes it easier to find the test, and the CLion integration of catch unit tests will work better

danpoe replied:
The test accesses some protected methods of `function_pointer_restrictionst`. So to access those we declare `get_function_pointer_by_name_restrictions_test()` as a friend (see at the top of the file).

thk123 reviewed (commented):
:no_entry_sign: need a test that assigns a fp on a line a few lines before the call

danpoe committed:
Refactor get_by_name_restriction()

- remove unnecessary loop
- make it return the restriction instead of inserting it into the map

danpoe committed:
Refactor error handling of function pointer restriction

danpoe commented:
@thk123 A test that assigns a function pointer a few lines before the call is not necessary. The reason we look for those assignments is because the `label_function_pointer_calls` pass inserts them, and it always does so immediately before the call. (I've also removed the loops that search for these assignments and replaced them by a check of just the previous instruction)

thk123 commented on src/goto-programs/restrict_function_pointers.cpp:
```diff
@@ -344,29 +382,112 @@ function_pointer_restrictionst::parse_function_pointer_restriction(
+  if(!can_cast_expr<dereference_exprt>(function))
+    return {};
+
+  auto const &function_pointer_call_site = to_symbol_expr(
```
:bulb: Even a comment here:

```C++
// function pointer guarateed to be a symbol by the function-pointer-label pass
```

Where function-pointer-label should be an actual file/folder/function that can be searched for!

thk123 reviewed (commented):
Looks good from a code perspective. 

Call me a Luddite, but even knowing the code doesn't rely on an assignment from a few lines before, I think it would be good to add the test, since not obvious to me that this is true. 

I also would feel better about the test without it having distinct regexes so we know it remains doing something

Other than that, lgtm :+1: thanks for addressing comments

smowton reviewed (approved):
Only reviewed the main changed, not the tests

thk123 reviewed (approved):
Dan pointed out that all the tests that don't have a local variable are testing what I was referring to.
